crates/metal-llvm/src/ty.rs

```diff
@@ -0,0 +1,40 @@
+use llvm_sys::core::{LLVMArrayType2, LLVMFunctionType, LLVMStructTypeInContext};
+use metal_mir::types::Type;
+
+use super::{get_types, CodeGenType};
+
+impl CodeGenType for Type {
+    fn codegen_type(
+        &self,
+        llvm: &crate::LLVMRefs,
+        module: &metal_mir::parcel::Module,
+    ) -> llvm_sys::prelude::LLVMTypeRef {
+        match self {
+            Self::Primitive(p) => p.codegen_type(llvm, module),
+            Self::Composite(c) => unsafe {
+                match *c.clone() {
+                    metal_mir::types::Composite::Tuple(t) => {
+                        let num_types = t.types.len();
+                        LLVMStructTypeInContext(
+                            llvm.ctx,
+                            get_types(llvm, module, t.types).as_mut_ptr(),
```

Comment:
The below reasoning is invalid as LLVM [copies](https://llvm.org/doxygen/Type_8cpp_source.html#l00447) the array data before returning the `StructType`. It remains because yes.

<details>
  <summary>Click to expand</summary>

To repeat my concern raised on Discord, i'm worried that this line can cause dead memory access.

`StructType`, returned by `LLVMStructTypeInContext`, inherits from `Type`, which contains a member called `ContainedTy` ([docs](https://llvm.org/doxygen/classllvm_1_1Type.html#a96d2037eb473457f3222e5cbe7ea22b5)). In case of `StructType`, this is where the parameter types are stored.

If we look at the [source](https://llvm.org/doxygen/IR_2Core_8cpp_source.html#l00789) of `LLVMStructTypeInContext`, we can trace the code to [`StructType::get`](https://llvm.org/doxygen/classllvm_1_1StructType.html#a18fc4545474c6ebb6f7c547f64f4fb31). One observation to be made already is that it takes [`ArrayRef`](https://llvm.org/doxygen/classllvm_1_1ArrayRef.html) as a parameter, which has this line in it's documentation:

> ..it [`ArrayRef`] is expected to be used in situations where the data resides in some other buffer, whose lifetime extends past that of the `ArrayRef`.

Continuing to trace the code further, we eventually find the code [calling `StructType::setBody`](https://llvm.org/doxygen/Type_8cpp_source.html#l00433), which finally assigns that `ArrayRef` to `ContainedTy`. This makes `StructType` valid only for as long as the array (`Vec`, in our case) passed to `LLVMStructTypeInContext` is.

Now, given Rust's rules around `Drop`, the `Vec` passed to LLVM will be deleted before the `.codegen_type` function returns, which makes the returned `StructType` reference deleted memory. UB!

To avoid this, we either need to use an allocator/`Vec` impl pair that only drops stuff when the allocator is dropped, or `std::mem::forget`/`Vec::leak` the `Vec`.

</details>